### ISSUE TITLE
`with` should return the value it was passed (if not given an object asset)

### ISSUE THREAD

**Issue by coyo-t** [YoYoGames/GameMaker-Bugs#6354]
### Is your feature request related to a problem?

Using `with` as an "initializer block" for new structs/instances is handy. however, you have to type the name twice (once for assigning a variable, and once for the with block)
this can result in very easy to miss mistakes like this. Can you spot it? because i didnt until literally just now
![image](https://github.com/YoYoGames/GameMaker-Bugs/assets/22027276/0548201a-668f-43a9-9f99-d8326805afcc)
while this error was *not* with a `with` block itself, it certainly could have been, and most certainly could have been avoided if `with` returned what it was given

### Describe the solution you'd like

something like
```gml
// struct
global.stone = with (blocks_register("stone", new Block()))
{
	colour = c_grey
	sound_type = global.block_soundtypes.tile
}

// instance
var thing = with (instance_create_depth(obj_thing, 0, 0, 0))
{
	// things here
}

// odd cases
var und = with undefined {
	// with block is not run, but will still return undefined
}

var no = with noone {
	// also does not run tmk, still returns noone
}

var obj = with obj_type {
	// unsure what exactly would it return, if anything? the obj asset? noone? undefined? -1?
	// probably should error if possible
}
```

### Describe alternatives you've considered

None that arent inconvienent or inefficient

if macros had arguments it would nearly be possible, but still not really. example:
```gml
#macro apply(o) method(o, function () {
#macro endapply ; return self; })();

var thing = apply (new Thing())
	colour = c_orange
	update_something()
endapply
```
this among other ideas ive had just dont work for a myriad of reasons, nonetheleast of which is gamemaker's lack of closures, making any solution that uses anon functions under the hood a practical no-go

### Additional context

im well aware of the separation of statements and expressions, however `with` is already a pretty strange language feature and i dont think making it operate as both a statement and a stronger version of `var a = (method(o, function () { return self; }))();` would make it any harder to learn how to use >_>

**Comment by backYard321** [comment]
I might be misunderstanding this, but couldn't you just do:

```gml
with blocks_register("stone", new Block())
{
	global.stone = self
	colour = c_grey
	sound_type = global.block_soundtypes.tile
}
```

**Comment by coyo-t** [comment]
this is true though very clunky looking and hard to follow
and wouldnt work very well outside of my specific ""global enum class"" setup for block types
i *have* done something similar in the past, like in `obj_game_manager`'s create event
```gml
with new Camera(...)
{
	other.cam = self
	// ...
}
```
but thats even worse to read

**Comment by backYard321** [comment]
Yeah idk i think this is kind of a cool idea! I feel like there are probably potential headaches but it would save the trouble of doing stuff like

```gml
var _wall = noone
with instance_place(x,y, oWall)
{
    collided = true
    other.speed = 0
    _wall = id
}
```
Instead it would just be
```gml
var _wall = with instance_place(x,y, oWall)
{
    collided = true
    other.speed = 0
}
```
and you wouldn't have to worry about taking into account `instance_place()` returning either an id or `noone`, you just use the function itself, and don't have to write the variable twice like in your examples.

I think where this maybe gets weird is returning values from functions, such as:

```gml
return with instance_place(x,y, oWall)
{
    collided = true
    other.speed = 0
}
```

But idk that seems like something most people could at least wrap their head around. So yeah, i think this is cool!

**Comment by rwkay** [comment]
`with` is a statement and not part of an expression - it does not and will not return anything.

It would be a fundamental change to GML to make these things part of an expression as that means you could do `with` statements in arguments to function or as the index to an array - these are nonsense things. What would the `with` return after it iterated over all the instances in a `with( all )` statement? 

This makes no sense at all.

**Comment by backYard321** [comment]
> `with` is a statement and not part of an expression - it does not and will not return anything.
> 
> It would be a fundamental change to GML to make these things part of an expression as that means you could do `with` statements in arguments to function or as the index to an array - these are nonsense things. What would the `with` return after it iterated over all the instances in a `with( all )` statement?
> 
> This makes no sense at all.

I mean it would just return `all`, and while i get that this is pretty different i think "nonsense" is an exaggeration, since alternate solutions to this specific issue are actively being discussed for GML V3.
